src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php

```diff
@@ -0,0 +1,136 @@
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Category\NameBuilder;
+
+use Cache;
+use PrestaShop\PrestaShop\Adapter\Category\Repository\CategoryRepository;
+use PrestaShop\PrestaShop\Core\Domain\Category\ValueObject\CategoryId;
+use PrestaShop\PrestaShop\Core\Domain\Language\ValueObject\LanguageId;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopId;
+
+/**
+ * Builds category display name to avoid confusing categories when they have identical names in multiple levels.
+ */
+class CategoryDisplayNameBuilder
+{
+    /**
+     * @var CategoryRepository
+     */
+    private $categoryRepository;
+
+    /**
+     * @param CategoryRepository $categoryRepository
+     */
+    public function __construct(
+        CategoryRepository $categoryRepository
+    ) {
+        $this->categoryRepository = $categoryRepository;
+    }
+
+    /**
+     *  If there are multiple categories with identical names, we want to be able to tell them apart,
+     *  so we use breadcrumb path instead of category name.
+     *  However, whole breadcrumb path would probably be too long, therefore not UX friendly.
+     *  Calculating "optimal" breadcrumb length seems too complex compared to the value it could bring.
+     *  So, we show one parent name and category name, as it is simple and should cover most cases.
+     *
+     * e.g. "Clothes > Women"
+     *
+     * @param CategoryId $categoryId
+     * @param string $categoryName
+     * @param ShopId $shopId
+     * @param LanguageId $languageId
+     * @param array|null $breadcrumbParts
+     * @param bool $useCache
+     *
+     * @return string
+     */
+    public function build(
+        CategoryId $categoryId,
+        string $categoryName,
+        ShopId $shopId,
+        LanguageId $languageId,
+        ?array $breadcrumbParts = null,
+        bool $useCache = true
+    ): string {
+        $duplicateNames = $this->getDuplicateNames($shopId, $languageId, $useCache);
+
+        if (!in_array($categoryName, $duplicateNames, true)) {
+            return $categoryName;
+        }
+
+        if (null === $breadcrumbParts) {
+            $breadcrumbParts = $this->categoryRepository->getBreadcrumbParts(
+                $categoryId,
+                $languageId
+            );
+        }
+
+        return implode(' > ', array_slice($breadcrumbParts, -2, 2));
```

Comment:
Isn't this code redundant with `CategoryRepository::getBreadcrumb(...)` ? Why not adding number of parents in `getBreadcrumb` arguments ?